designer/client/inline-condition-model.js

```diff
@@ -0,0 +1,327 @@
+  }
+
+  _autoGroupDefs (conditions) {
+    const orPositions = []
+    conditions.forEach((condition, index) => {
+      if (condition.getCoordinator() === coordinators.OR) {
+        orPositions.push(index)
+      }
+    })
+    const hasAnd = !!conditions.find(condition => condition.getCoordinator() === coordinators.AND)
+    const hasOr = orPositions.length > 0
+    if (hasAnd && hasOr) {
+      let start = 0
+      const groupDefs = []
+      orPositions.forEach((position, index) => {
+        if (start < position - 1) {
+          groupDefs.push(new GroupDef(start, position - 1))
+        }
+        const thisIsTheLastOr = orPositions.length === index + 1
+        const thereAreMoreConditions = conditions.length - 1 > position
+        if (thisIsTheLastOr && thereAreMoreConditions) {
+          groupDefs.push(new GroupDef(position, conditions.length - 1))
+        }
+        start = position
+      })
+      return groupDefs
+    }
+    return []
+  }
+
+  clone () {
+    const toReturn = new ConditionsModel()
+    toReturn.groupedConditions = this.groupedConditions.map(it => it.clone())
+    toReturn.userGroupedConditions = this.userGroupedConditions.map(it => it.clone())
+    return toReturn
+  }
+}
+
+export class GroupDef {
+  constructor (first, last) {
+    if (typeof first !== 'number' || typeof last !== 'number') {
+      throw Error(`Cannot construct a group from ${first} and ${last}`)
+    } else if (first >= last) {
+      throw Error(`Last (${last}) must be greater than first (${first})`)
+    }
+    this.first = first
+    this.last = last
+  }
+
+  contains (index) {
+    return this.first <= index && this.last >= index
+  }
+
+  startsWith (index) {
+    return this.first === index
+  }
+
+  applyTo (conditions) {
+    return [...conditions].splice(this.first, (this.last - this.first) + 1)
+  }
+}
+
+class ConditionGroup {
+  constructor (conditions) {
+    if (!Array.isArray(conditions) || conditions.length < 2) {
+      throw Error(`Cannot construct a condition group from ${conditions}`)
+    }
+    this.conditions = conditions
+  }
+
+  toPresentationString () {
+    const copy = [...this.conditions]
+    copy.splice(0, 1)
+    return `${this.coordinatorString()}${this.conditionString()}`
+  }
+
+  coordinatorString () {
+    return this.conditions[0].coordinatorString()
+  }
+
+  conditionString () {
+    const copy = [...this.conditions]
+    copy.splice(0, 1)
+    return `(${this.conditions[0].conditionString()} ${copy.map(condition => condition.toPresentationString()).join(' ')})`
+  }
+
+  asFirstCondition () {
+    this.conditions[0].asFirstCondition()
+    return this
+  }
+
+  getCoordinator () {
+    return this.conditions[0].getCoordinator()
+  }
+
+  setCoordinator (coordinator) {
+    this.conditions[0].setCoordinator(coordinator)
+  }
+
+  isGroup () {
+    return true
+  }
+
+  clone () {
+    return new ConditionGroup(this.conditions.map(condition => condition.clone()))
+  }
+}
+
+export class Condition {
```

Comment:
No need to action this, it's fine for now, `ConditionGroup` should be subclass of `Condition` since their 'interfaces' are the same

Reply:
Can I do this? The constructor signatures are different so it'd involve creating an invalid condition?

Reply:
ahhh I see! lets hope es gets interfaces/protocols soon then 🤓